api/inc/register_gateway.h

```diff
@@ -17,105 +17,146 @@
+/** Both `uvisor_read` and `uvisor_write` can have 2 additional input arguments -
+ *  mask and operation. These macros select the right argument from
+ *  `__VA_ARGS__` and provide a sensible default if none is provided.
+ * @internal
+ */
+#define SELECT_MASK(...)       __UVISOR_MACRO_SELECT(_0, ##__VA_ARGS__, , , _MASK , , _DEFAULT_NULL)(__VA_ARGS__)
+#define SELECT_READ_OP(...)    __UVISOR_MACRO_SELECT(_0, ##__VA_ARGS__, , , _OP , , _DEFAULT_READ)(__VA_ARGS__)
+#define SELECT_WRITE_OP(...)   __UVISOR_MACRO_SELECT(_0, ##__VA_ARGS__, , , _OP , , _DEFAULT_WRITE)(__VA_ARGS__)
+#define _MASK(mask, operation) mask
+#define _OP(mask, operation)   operation
+#define _DEFAULT_NULL(...)     0
+#define _DEFAULT_READ(...)     UVISOR_RGW_OP_READ
+#define _DEFAULT_WRITE(...)    UVISOR_RGW_OP_WRITE
+
+/** Get the offset of a struct member.
+ * @internal
+ */
+#define OFFSETOF(type, member) ((uint32_t) (&(((type *)(0))->member)))
+
+/** Generate the opcode of the 16-bit Thumb-2 16-bit T2 encoding of the branch
+ *  instruction.
+ * @internal
+ * @note The branch instruction is encoded according to the Thumb-2 immediate
+ * encoding rules:
+ *     <instr_addr>: B.N <label>
+ *     imm = (<label>_addr - PC) / 2
+ * Where:
+ *     PC = <instr>_addr + 4
+ * The +4 is to account for the pipelined PC at the time of the branch
+ * instruction. See ARM DDI 0403E.b page A4-102 for more details.
+ * @param instr[in] Address of the branch instruction
+ * @param label[in] Address of the label
+ * @returns the 16-bit encoding of the B.N <label> instruction.
+ */
+#define BRANCH_OPCODE(instr, label) \
+    (uint16_t) (0xE000 | (uint8_t) ((((uint32_t) (label) - (uint32_t) (instr + 4)) / 2) & 0xFF))
 
-#define __UVISOR_REGISTER_GATEWAY_METADATA(src_box, ...) \
-     __UVISOR_MACRO_SELECT(_0, ##__VA_ARGS__, __UVISOR_REGISTER_GATEWAY_METADATA4, \
-                                              __UVISOR_REGISTER_GATEWAY_METADATA3, \
-                                              __UVISOR_REGISTER_GATEWAY_METADATA2, \
-                                              __UVISOR_REGISTER_GATEWAY_METADATA1, \
-                                              /* No macro for 0 args */          )(src_box, ##__VA_ARGS__)
+/** `BX LR` encoding
+ * @internal
+ */
+#define BXLR                           ((uint16_t) 0x4770)
 
-/* Rregister-level gateway - Read */
-/* FIXME: Currently only a hardcoded 32bit constant can be used for the addr
- *        field. */
-#define uvisor_read(src_box, ...) \
+/** Register Gateway - Read operation
+ *
+ * This macro provides a variadic API to perform 32-bit read operations on
+ * restricted registers. Such accesses are assembled into a read-only flash
+ * structure that is read and validated by uVisor before performing the
+ * operation.
+ *
+ * The following APIs can be used:
+ * void uvisor_read(src_box, address);
+ * void uvisor_read(src_box, address, operation, mask);
+ *
+ * @warning These APIs currently only support link-time known values. This means
```

Comment:
What does this warning mean for the `uvisor_read` operation? The following example is for `uvisor_write`. Consider either removing the warning to making the example illustrate what the warning means.